File: 49-group-anagrams/49-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        
        for(String s : strs) {
            String enc = encode(s);
            List<String> list = map.computeIfAbsent(enc, e -> new ArrayList<>());
            list.add(s);
        }
        
        return new ArrayList<>(map.values());
    }
    
    String encode(String str) {
        int[] bucket = new int[26];
        
        for(int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            bucket[c - 'a']++;
        }
        
        StringBuilder sb = new StringBuilder();
        
        for(int i = 0; i < 26; i++) {
            sb.append(bucket[i]);
            sb.append(',');
        }
        
        return sb.toString();
    }
}

File: 49-group-anagrams/49-group-anagrams.scala
import scala.collection.mutable._

object Solution {
    def groupAnagrams(strs: Array[String]): List[List[String]] = {
        val map: Map[String, ListBuffer[String]] = Map()

        strs.foreach { str =>
            val encoded = encode(str)
            val list = map.get(encoded).getOrElse(ListBuffer[String]())
            list += str
            map += (encoded -> list)
        }

        val res: ListBuffer[List[String]] = ListBuffer()

        map.foreach { kv => 
            res += kv._2.toList
        }

        res.toList
    }

    private def toInt(char: Char): Int = 
        char.asInstanceOf[Int] - 'a'.asInstanceOf[Int]
    
    private def encode(str: String): String = {
        val arr: Array[Int] = Array.fill(26)(0)
        var res = ""

        str.foreach { c => 
            arr(toInt(c)) += 1
        }
        arr.foreach { n => 
            res += n + ";"
        }

        res
    }
}